File: tests/VDG.Core.Tests/PipelineTests.fs
namespace VDG.Core.Tests

open Xunit
open VDG.Core

module PipelineTests =

    [<Fact>]
    let ``buildCommands returns nodes then connectors`` () =
        let nA = { Id="A"; Label="Alpha"; Children=[]; Parent=None }
        let nB = { Id="B"; Label="Beta";  Children=[]; Parent=None }
        let c1 = { From="A"; To="B"; Label=None }

        let cmds = Pipeline.buildCommands [ nA; nB ] [ c1 ]
        Assert.Equal(3, cmds.Length)

        match cmds.[0], cmds.[1], cmds.[2] with
        | AddNode n1, AddNode n2, AddConnector c ->
            Assert.Equal("A", n1.Id)
            Assert.Equal("B", n2.Id)
            Assert.Equal("A", c.From)
            Assert.Equal("B", c.To)
        | _ -> failwith "Unexpected command ordering"

File: tests/p9-11_tests/PipelineTests.cs
using System.Collections.Generic;
using NSubstitute;
using VDG.Core;
using Xunit;

namespace VDG.Tests.P9_11
{
    public class PipelineTests
    {
        [Fact]
        public void Aggregates_Items_And_Connections_Into_Commands()
        {
            var p1 = Substitute.For<IMapProvider>();
            p1.GetItems().Returns(new List<DiagramItem> { new DiagramItem("1","T","One",0,0) });
            p1.GetConnections().Returns(new List<DiagramConnection>());

            var p2 = Substitute.For<IMapProvider>();
            p2.GetItems().Returns(new List<DiagramItem> { new DiagramItem("2","T","Two",0,0) });
            p2.GetConnections().Returns(new List<DiagramConnection> { new DiagramConnection("1","2","line") });

            var pipe = new Pipeline(new[] { p1, p2 }, new GridLayoutAlgorithm());
            var cmds = pipe.BuildCommands();

            Assert.Equal(3, cmds.Count); // 2 shapes + 1 connector
        }
    }
}